### PR TITLE
Fix: add recursive inode/dentry invalidation and API smoke test

### LINKED ISSUE #1694: invalidate dentry/inode cache after 255 nydusd mounts
After performing more than 255 sub-mount operations on nydusd (even if unmounting follows mounting), there is a probability that file system corruption will occur when users attempt to mount again.

This issue manifests as accessing files or content from previous mount points. The root cause lies in the fuse-backend-rs library, which uses a u8 type for pseudo inode indexing. Additionally, after unmounting sub-mount points, the library does not call FUSE's invalidate dentry cache or invalidate inode cache. As a result, after exceeding 255 mounts, new mountpoints may reuse the pseudo inode of sub-mount points due to kernel cache not being cleared, leading to file system inconsistencies.

### PR DESCRIPTION
**Pull request by sidneychang**
## Overview
Performing more than 255 sub-mount operations on nydusd can lead to file system corruption, even if each mount is properly unmounted. The corruption occurs because fuse-backend-rs uses a u8 type for pseudo inode indexing, and after unmounting, it does not invalidate the FUSE dentry or inode caches. As a result, kernel caches may retain stale pseudo inodes. Once the 255 pseudo inode limit is exceeded, new mount points can reuse old pseudo inodes, causing users to access files or content from previous mounts, leading to inconsistencies.
## Related Issues
Fix #1694 

## Change Details
- Before unmounting a RAFS filesystem, recursively invalidate all inodes and dentries to prevent stale kernel caches when using shared mounts.
- Added a new smoke test `TestSubMountCache` to mount/unmount sub-mounts multiple times (300 iterations) and verify file, directory, symlink.

## Test Results

In the smoke test, the first mount uses a “thin” filesystem containing fewer files, while the next 255 mounts use a full filesystem with more files.

Before this patch, on the 256th mount, accessing files that exist in the full filesystem but not in the thin filesystem could fail:
```
[2025-08-28 13:39:12.597203 +08:00] INFO RAFS filesystem imported
[2025-08-28 13:39:12.597339 +08:00] INFO Rafs filesystem mounted at /mount-255
[2025-08-28 13:39:12.597602 +08:00] ERROR fuse: reply error header OutHeader { len: 16, error: -5, unique: 72202 }, error Custom { kind: NotADirectory, error: "Not a directory (os error 20)" }
api_test.go:263: 
        Error Trace:    /root/nydus/smoke/tests/api_test.go:263
        Error:          Received unexpected error:
                        stat /tmp/nydus-smoke-1919054383/mnt/mount-255/dir-1/file-1: input/output error
        Test:           TestAPI/TestSubMountCache
        Messages:       stat file-1 failed


```
This happens because the pseudo inode cache from the first mount (thin filesystem) is reused in the 256th mount, and the kernel cache has not been invalidated. As a result, the kernel sees stale pseudo inodes, causing read errors when accessing files that exist in the full filesystem but not in the thin one.

After applying this patch, the test passes completely, as recursive inode/dentry invalidation ensures that kernel caches are cleared and pseudo inode reuse no longer causes inconsistencies.
```
[2025-08-28 13:59:08.047650 +08:00] INFO State machine(pid=1116461): from Running to Ready, input [Stop], output [Some(TerminateService)]
[2025-08-28 13:59:08.047787 +08:00] INFO State machine(pid=1116461): from Ready to Die, input [Stop], output [Some(Umount)]
[2025-08-28 13:59:08.047835 +08:00] INFO state_machine thread exits
--- PASS: TestAPI (0.00s)
    --- PASS: TestAPI/TestSubMountCache (9.30s)
PASS
```

## Change Type
_Please select the type of change your pull request relates to:_
- [x] Bug Fix
- [ ] Feature Addition
- [ ] Documentation Update
- [ ] Code Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe)

## Self-Checklist
_Before submitting a pull request, please ensure you have completed the following:_
- [ ] I have run a code style check and addressed any warnings/errors.
- [ ] I have added appropriate comments to my code (if applicable).
- [ ] I have updated the documentation (if applicable).
- [ ] I have written appropriate unit tests.

### CODE REVIEW

**Comment by sidneychang** [comment]
Due to modifications in fuse-backend-rs, the project is now using version v0.13.0. The code constructs FuseDevWriter directly from a file descriptor (raw_fd) to work around current limitations.
There are two known issues in fuse-backend-rs:
**with_writer**: Switching to with_writer could simplify the code, but it requires additional macOS support in the library. Without that, CI builds on macOS would fail.
**unsafe usage**: There are two places in the library where certain calls need to be wrapped in unsafe blocks.
For now, we continue using the raw_fd approach, accepting that CI may fail, and will revisit these issues when we decide on the proper modifications to the library.

**Comment by sidneychang** [comment]
I am currently using my own fork of fuse-backend-rs to verify whether the CI can pass. Once the corresponding commits are merged into the upstream fuse-backend-rs repository and a release is published, I will switch back.

File: service/src/fusedev.rs [hunk 1]
```diff
@@ -198,6 +205,93 @@ impl FusedevFsService {
...
+        stack.push((
+            ROOT_PARENT_INO,
+            path.trim_start_matches('/').to_string(),
+            inode.clone(),
+            false,
+        ));
+
+        while let Some((parent_kernel_ino, cur_name, cur_inode, visited)) = stack.pop() {
+            // Convert rafs inode to inode id used by kernel
+            let cur_kernel_ino = ((fs_idx as u64) << FS_IDX_SHIFT) | cur_inode.ino();
```
Comment (imeoer) [inline service/src/fusedev.rs]: May be use `Vfs::convert_inode` here (not require for this PR).

File: service/src/fusedev.rs [hunk 2]
```diff
@@ -198,6 +205,93 @@ impl FusedevFsService {
 
         inflight_op
     }
+    fn walk_and_notify_invalidation(
```
Comment (imeoer) [inline service/src/fusedev.rs]: Can we reuse `RafsInode::walk_children_inodes` here?

File: service/src/fusedev.rs [hunk 3]
```diff
@@ -237,6 +331,31 @@ impl FsService for FusedevFsService {
         }
     }
 
+    fn umount(&self, cmd: FsBackendUmountCmd) -> NydusResult<()> {
```
Comment (imeoer) [inline service/src/fusedev.rs]: Can we call `service/src/fusedev.rs -> walk_and_notify_invalidation` in the `service/src/fs_service.rs -> umount`? So that we can reuse the `umount` codes.

File: smoke/tests/api_test.go [hunk 4]
```diff
@@ -193,6 +193,109 @@ func (a *APIV1TestSuite) TestPrefetch(t *testing.T) {
...
+		EnablePrefetch:  ctx.Runtime.EnablePrefetch,
+		DigestValidate:  false,
+		AmplifyIO:       ctx.Runtime.AmplifyIO,
+	}
+
+	// mount prefix where the daemon actually mounts sub-mounts
+	mountPrefix := ctx.Env.WorkDir + "/mnt"
+
+	// iterate 256 times; first mount uses thinlower (without some files), subsequent mounts use Lower.
+	for i := 0; i < 256; i++ {
```
Comment (imeoer) [inline smoke/tests/api_test.go]: It's better to generate different file names and content (e.g., the `texture.MakeMatrixLayer` method), mount exceed 255 count, and verify the fs.

File: service/src/fusedev.rs (hunk 3 above)
Comment (sidneychang) [reply service/src/fusedev.rs]: I’m not sure whether the kernel cache should always be invalidated in all scenarios. For virtiofs, the upstream PR [fuse-backend-rs#152](https://github.com/cloud-hypervisor/fuse-backend-rs/pull/152)
 notes that VFS keeps pseudo inodes on umount to support hot upgrade and quick recovery, which suggests that retaining the kernel dentry/inode cache might also be intentional. For normal fuse, however, not clearing the cache exposes invalid directories on the host, so explicit invalidation is required. If we decide that invalidating the cache is necessary in all cases, I can move the logic into fs_service::umount so it’s handled consistently.

> Can we call `service/src/fusedev.rs -> walk_and_notify_invalidation` in the `service/src/fs_service.rs -> umount`? So that we can reuse the `umount` codes.

File: service/src/fusedev.rs (hunk 2 above)
Comment (sidneychang) [inline service/src/fusedev.rs:208]: walk_children_inodes is already used here to collect the children. The reason I implemented a manual DFS stack is that in the v5 metadata implementation, the handler doesn’t provide an Arc<dyn RafsInode>, so recursion can’t be done directly. If we only need to support v6 (which passes the inode to the handler), then yes, I can simplify this and reuse walk_children_inodes directly for recursion.

File: smoke/tests/api_test.go (hunk 4 above)
Comment (sidneychang) [reply smoke/tests/api_test.go:244]: Thanks for pointing this out — the current test only repeats with fixed file names and content. I’ll switch to MakeMatrixLayer to generate varied files and mount more than 255 times to better cover cache and boundary cases.

File: service/src/fusedev.rs (hunk 2 above)
Comment (sidneychang) [reply service/src/fusedev.rs]: walk_children_inodes is already used here to collect the children. The reason I implemented a manual DFS stack is that in the v5 metadata implementation, the handler doesn’t provide an Arc, so recursion can’t be done directly. If we only need to support v6 (which passes the inode to the handler), then yes, I can simplify this and reuse walk_children_inodes directly for recursion.

File: service/src/fs_service.rs [hunk 5]
```diff
@@ -32,6 +33,8 @@ use versionize_derive::Versionize;
 use crate::upgrade::UpgradeManager;
 use crate::{Error, FsBackendDescriptor, FsBackendType, Result};
 
+const ROOT_PARENT_INO: u64 = 1;
```
Comment (imeoer) [inline service/src/fs_service.rs]: Mey be use `fuse_backend_rs::abi::fuse_abi::ROOT_ID`.

File: service/src/fusedev.rs [hunk 6]
```diff
@@ -335,6 +339,65 @@ impl FsService for FusedevFsService {
...
+            cur_inode.walk_children_inodes(0, &mut handler)?;
+        }
+
+        // === Post-order: invalidate cache of the current node ===
+        let cstr_name = CString::new(cur_name).map_err(|_| eother!("invalid file name"))?;
+        // Invalidate inode cache
+        self.session.lock().unwrap().with_writer(|writer| {
+            if let Err(e) = self.server.notify_inval_inode(writer, cur_kernel_ino, 0, 0) {
+                warn!("notify_inval_inode failed: {} {:?}", cur_name, e);
+            }
```
Comment (imeoer) [inline service/src/fusedev.rs:381]: Can we also put `notify_inval_entry` here.

File: smoke/tests/api_test.go [hunk 7]
```diff
@@ -192,6 +192,80 @@ func (a *APIV1TestSuite) TestPrefetch(t *testing.T) {
...
+		curCfg := childTpl
+		curCfg.BootstrapPath = bootstrap
+		curCfg.MountPath = childTpl.MountPath + fmt.Sprintf("-%d", i)
+
+		err = nydusd.MountByAPI(curCfg)
+		require.NoError(t, err, "failed to mount by API at iteration %d", i)
+
+		fullMount := filepath.Join(mountPrefix, curCfg.MountPath)
+
+		// Verify file contents
```
Comment (imeoer) [inline smoke/tests/api_test.go]: It seems the check is unnecessary, since `nydusd.VerifyByPath` will also verify the file's metadata and content.

**Review by imeoer — COMMENTED** [review]
Thanks, others LGTM!

File: service/src/fusedev.rs (hunk 6 above)
Comment (sidneychang) [reply service/src/fusedev.rs:381]: Since both notify_inval_inode and notify_inval_entry require writer as a mutable borrow, we cannot perform two write operations within the same with_writer closure. Each notify function requires exclusive access to writer, so we must call with_writer twice.
```
pub fn notify_inval_inode<S: BitmapSlice>(
    &self,
    mut w: FuseDevWriter<'_, S>,  // mutable borrow
    ...
) 

pub fn notify_inval_entry<S: BitmapSlice>(
    &self,
    mut w: FuseDevWriter<'_, S>,  // mutable borrow
    ...
) -
```

**Review by imeoer — APPROVED** [review]
Thanks, LGTM!

**Comment by imeoer** [comment]
The broken CI will be resolved next PR.
